crates/chaintsn/src/epoch.rs

```diff
@@ -0,0 +1,381 @@
+//! Epoch check-in logic.
+
+use rand_core::{RngCore, SeedableRng};
+use strata_primitives::params::RollupParams;
+use strata_state::{
+    block::L1Segment,
+    bridge_ops::{DepositIntent, WithdrawalIntent},
+    bridge_state::*,
+    id::L2BlockId,
+    l1::{L1BlockId, L1MaturationEntry},
+    state_op::*,
+    tx::ProtocolOperation,
+};
+
+use crate::{errors::TsnError, slot_rng::SlotRng};
+
+/// Rollup epoch-level input.
+pub struct EpochData {
+    final_l2_blockid: L2BlockId,
+    l1_segment: L1Segment,
+    // TODO deposits, DA, checkpoints
+}
+
+/// Performs the once-per-epoch updates we make to a block.
+///
+/// This is invoked after the core block STF on the last block of an epoch to
+/// perform checkins with the L1 state.
+pub fn process_epoch(
+    state: &mut StateCache,
+    epoch_data: EpochData,
+    params: &RollupParams,
+) -> Result<(), TsnError> {
+    // FIXME make this actually init correctly
+    let mut rng = SlotRng::from_seed(*epoch_data.final_l2_blockid.as_ref());
+
+    // Assign withdrawals to deposits.
+    process_l1_segment(state, &epoch_data.l1_segment, params)?;
+    process_deposit_updates(state, &mut rng, params)?;
+
+    Ok(())
+}
+
+/// Update our view of the L1 state, playing out downstream changes from that.
+fn process_l1_segment(
+    state: &mut StateCache,
+    l1seg: &L1Segment,
+    params: &RollupParams,
+) -> Result<(), TsnError> {
+    // Accept new blocks, comparing the tip against the current to figure out if
+    // we need to do a reorg.
+    // FIXME this should actually check PoW, it just does it based on block heights
+    if !l1seg.new_payloads().is_empty() {
+        // Validate the new blocks actually extend the tip.  This is what we have to tweak to make
+        // more complicated to check the PoW.
+        let new_tip_block = l1seg.new_payloads().last().unwrap();
+        let new_tip_blkid = new_tip_block.record().blkid();
+        let new_tip_height = new_tip_block.idx();
+        let first_new_block_height = new_tip_height - l1seg.new_payloads().len() as u64 + 1;
+        let cur_safe_height = state.l1_safe_height();
+
+        // Check that the new chain is actually longer, if it's shorter then we didn't do anything.
+        // TODO This probably needs to be adjusted for PoW.
+        if new_tip_height < cur_safe_height {
+            return Err(TsnError::L1SegNotExtend);
+        }
+
+        // TODO make sure that the block hashes all connect up sensibly.
+
+        // Collect the indexes of current operators.
+        let op_idxs: Vec<_> = state.epoch_state().operator_table().indices().collect();
+
+        for e in l1seg.new_payloads() {
+            // TODO maybe consolidate these to have ops for each tx
+
+            for tx in e.deposit_update_txs() {
+                if let ProtocolOperation::Deposit(deposit_info) = tx.tx().protocol_operation() {
```

Comment:
Asking for deposit_update_txs and then checking if the pattern match and `.tx().protocol_operation()` look a bit weird, the interface could be simpler I guess, just get the deposit infos from payload.